### ISSUE TITLE
enum and define with the same name collide

### ISSUE THREAD

**Issue by deweerdt** [rust-lang-nursery/rust-bindgen#687]
### Input C/C++ Header

```C
enum {
        A =
#define A 0
                A,
} anenum;
```

### Bindgen Invokation

<!-- Place either the `bindgen::Builder` or the command line flags used here. -->

```Rust
bindgen::Builder::default()
    .header("input.h")
    .generate()
    .unwrap()
```


### Actual Results


```
       Fresh bindgen v0.23.1                                                                                                                                                                                                                                                                                                        Compiling math_dot_h v0.1.0 (file:///home/def/p/rust/math_dot_h)
     Running `rustc --crate-name math_dot_h src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=93892b9e7e7b815c -C extra-filename=-93892b9e7e7b815c --out-dir /home/def/p/rust/math_dot_h/target/debug/deps -L dependency=/home/def/p/rust/math_dot_h/target/debug/deps`
error[E0428]: a value named `A` has already been defined in this module
 --> /home/def/p/rust/math_dot_h/target/debug/build/math_dot_h-7f0a85db403a8610/out/bindings.rs:4:1
  |
3 | pub const A: ::std::os::raw::c_uint = 0;
  | ---------------------------------------- previous definition of `A` here
4 | pub const A: _bindgen_ty_1 = _bindgen_ty_1::A;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `A` already defined

error: aborting due to previous error

error: Could not compile `math_dot_h`.

Caused by:
  process didn't exit successfully: `rustc --crate-name math_dot_h src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=93892b9e7e7b815c -C extra-filename=-93892b9e7e7b815c --out-dir /home/def/p/rust/math_dot_h/target/debug/deps -L dependency=/home/def/p/rust/math_dot_h/target/debug/deps` (exit code: 101)
```

### Expected Results

No compilation error, both the enum and the define have their own name space.

This is problematic IRL because `math.h` uses this construct to define `FP_*` constants:
https://sourceware.org/git/?p=glibc.git;a=blob;f=math/math.h;h=cfaed0ed98013830421afdf1ba6f13c01702831d;hb=HEAD#l318

```c
enum                                                                                                                                                                                                                                                                                                                               {                                                                                                                                                                                                                                                                                                                                  FP_NAN =                                                                                                                                                                                                                                                                                                                     # define FP_NAN 0                                                                                                                                                                                                                                                                                                                      FP_NAN,                                                                                                                                                                                                                                                                                                                        FP_INFINITE =                                                                                                                                                                                                                                                                                                                # define FP_INFINITE 1                                                                                                                                                                                                                                                                                                                 FP_INFINITE,
    FP_ZERO =
# define FP_ZERO 2
      FP_ZERO,
    FP_SUBNORMAL =
# define FP_SUBNORMAL 3
      FP_SUBNORMAL,
    FP_NORMAL =
# define FP_NORMAL 4
      FP_NORMAL
  };
```

### `RUST_LOG=bindgen` Output

<details>

```
   Compiling math_dot_h v0.1.0 (file:///home/def/p/rust/math_dot_h)
error[E0428]: a value named `A` has already been defined in this module
 --> /home/def/p/rust/math_dot_h/target/debug/build/math_dot_h-7f0a85db403a8610/out/bindings.rs:4:1
  |
3 | pub const A: ::std::os::raw::c_uint = 0;
  | ---------------------------------------- previous definition of `A` here
4 | pub const A: _bindgen_ty_1 = _bindgen_ty_1::A;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `A` already defined

error: aborting due to previous error

error: Could not compile `math_dot_h`.

To learn more, run the command again with --verbose.
```

</details>

**Comment by emilio** [comment]
That's... nasty...

What would be a good solution for this? I can allow you to ignore a given macro by name with a new method in `ParseCallbacks`, would that work for you?

**Comment by deweerdt** [comment]
Thanks for your quick response!

If i understand correctly, you're suggesting that i return `Ignore` for the definitions I know will also be used as defines? I believe that will work for me, yes.

In general, it feels like the two definitions should belong to two different namespaces, but i don't know enough about `bindgen` to suggest anything beyond offering my opinion.

**Comment by emilio** [comment]
Yeah, it's a kind of unfortunate trade-off between ease to use and features... In general bindgen doesn't need to do a lot of deduplicating at the root level because rust allows to differentiate `struct foo {}` from `fn foo`, and similar stuff.

We could either mangle stuff like `enum_foo`, `struct_bar`, etc... but that'd be really annoying, specially for C++ users... Oh well.

**Comment by retrry** [comment]
I've started seeing this behavior on bindgen 0.26. While bindgen 0.25 generated code without errors. Bindgen 0.26 started giving me errors:
```
error[E0428]: the name `FP_NAN` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2820:1
     |
491  | pub const FP_NAN: libc::c_int = 0;
     | ---------------------------------- previous definition of the value `FP_NAN` here
...
2820 | pub const FP_NAN: _bindgen_ty_1 = _bindgen_ty_1::FP_NAN;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NAN` redefined here
     |
     = note: `FP_NAN` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_INFINITE` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2821:1
     |
492  | pub const FP_INFINITE: libc::c_int = 1;
     | --------------------------------------- previous definition of the value `FP_INFINITE` here
...
2821 | pub const FP_INFINITE: _bindgen_ty_1 = _bindgen_ty_1::FP_INFINITE;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_INFINITE` redefined here
     |
     = note: `FP_INFINITE` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_ZERO` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2822:1
     |
493  | pub const FP_ZERO: libc::c_int = 2;
     | ----------------------------------- previous definition of the value `FP_ZERO` here
...
2822 | pub const FP_ZERO: _bindgen_ty_1 = _bindgen_ty_1::FP_ZERO;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_ZERO` redefined here
     |
     = note: `FP_ZERO` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_SUBNORMAL` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2823:1
     |
494  | pub const FP_SUBNORMAL: libc::c_int = 3;
     | ---------------------------------------- previous definition of the value `FP_SUBNORMAL` here
...
2823 | pub const FP_SUBNORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_SUBNORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_SUBNORMAL` redefined here
     |
     = note: `FP_SUBNORMAL` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_NORMAL` is defined multiple times
    --> /home/target/debug/build/ffmpeg-sys-026db31dcba0e79a/out/bindings.rs:2824:1
     |
495  | pub const FP_NORMAL: libc::c_int = 4;
     | ------------------------------------- previous definition of the value `FP_NORMAL` here
...
2824 | pub const FP_NORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_NORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NORMAL` redefined here
     |
     = note: `FP_NORMAL` must be defined only once in the value namespace of this module

error: aborting due to 5 previous errors

error: Could not compile `ffmpeg-sys`.
```

As a workaround I'm just ignoring those types:
```
        .hide_type("FP_NAN")
        .hide_type("FP_INFINITE")
        .hide_type("FP_ZERO")
        .hide_type("FP_SUBNORMAL")
        .hide_type("FP_NORMAL")
```

**Comment by gentoo90** [comment]
I've encountered the same error with `bindgen-0.25.5` and `clang-4.0.0` on linux.
Crate [`magick-rust`](https://github.com/nlfiedler/magick-rust).
```
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4825:1
     |
747  | pub const FP_NAN: libc::c_uint = 0;
     | ----------------------------------- previous definition of the value `FP_NAN` here
...
4825 | pub const FP_NAN: _bindgen_ty_1 = _bindgen_ty_1::FP_NAN;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NAN` redefined here
     |
     = note: `FP_NAN` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_INFINITE` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4826:1
     |
748  | pub const FP_INFINITE: libc::c_uint = 1;
     | ---------------------------------------- previous definition of the value `FP_INFINITE` here
...
4826 | pub const FP_INFINITE: _bindgen_ty_1 = _bindgen_ty_1::FP_INFINITE;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_INFINITE` redefined here
     |
     = note: `FP_INFINITE` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_ZERO` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4827:1
     |
749  | pub const FP_ZERO: libc::c_uint = 2;
     | ------------------------------------ previous definition of the value `FP_ZERO` here
...
4827 | pub const FP_ZERO: _bindgen_ty_1 = _bindgen_ty_1::FP_ZERO;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_ZERO` redefined here
     |
     = note: `FP_ZERO` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_SUBNORMAL` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4828:1
     |
750  | pub const FP_SUBNORMAL: libc::c_uint = 3;
     | ----------------------------------------- previous definition of the value `FP_SUBNORMAL` here
...
4828 | pub const FP_SUBNORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_SUBNORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_SUBNORMAL` redefined here
     |
     = note: `FP_SUBNORMAL` must be defined only once in the value namespace of this module

error[E0428]: the name `FP_NORMAL` is defined multiple times
    --> /home/user/magick-rust/target/debug/build/magick_rust-9bda580053656820/out/bindings.rs:4829:1
     |
751  | pub const FP_NORMAL: libc::c_uint = 4;
     | -------------------------------------- previous definition of the value `FP_NORMAL` here
...
4829 | pub const FP_NORMAL: _bindgen_ty_1 = _bindgen_ty_1::FP_NORMAL;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_NORMAL` redefined here
     |
     = note: `FP_NORMAL` must be defined only once in the value namespace of this module

```

**Comment by spacekookie** [comment]
Is there any update or progress on fixing this issue? I'm only affected by it via an external dependency so manually tweaking bindgen settings isn't really an option for me. And this is currently blocking a project 😞

**Comment by fitzgen** [comment]
I could imagine the `BindgenContext` carrying a map of `name -> counter` and writing a new pass over the IR right before codegen that bumps the appropriate entry's counter for each definition of any kind, and then appending the counter to the definition's name when its not the first time we've seen that name.

Essentially, moving our method/function overloading strategy up and out of codegen and use it for all definitions.

This should also solve https://github.com/rust-lang-nursery/rust-bindgen/issues/840

If anyone is interested in trying their hand at this, I can help mentor (cc @spacekookie )

**Comment by spacekookie** [comment]
Hey @fitzgen, I mean...I'd love to have a try at fixing this though I've only been writing Rust for ~2 months now and I'm rather green, especially when it comes to _large_ Rust projects. If you're offering to mentor though, maybe get in touch with me via [twitter](https://twitter.com/spacekookie) or [e-mail](mailto:kookie@spacekookie.de)?

**Comment by emilio** [comment]
I just noticed that there's a way to properly handling this hiding the enum instead of the constant with `ParseCallbacks::enum_variant_behavior`. Will add the symmetric API for macros, and then I think we can call it fixed. Of course we may want to expose that to the cli somehow, but that can be another issue.

**Comment by emilio** [comment]
Yes, but now you have a way to avoid them, using `MacroParsingBehavior` to avoid the relevant macros.

**Comment by xxks-kkk** [comment]
Hello @emilio ,

Can you please shed some lights on how to use `MacroParsingBehavior`? I'm new to Rust and bindgen and I just followed the tutorial to generate the bindings. I got the two following definitions

```
pub const IPPORT_RESERVED: ::std::os::raw::c_uint = 1024;
pub const IPPORT_RESERVED: _bindgen_ty_7 = _bindgen_ty_7::IPPORT_RESERVED;
```

However, none of them are part of my orginal C code. I'm wondering how can I fix this issue?

Thanks much!

**Comment by emilio** [comment]
You can either add a `ParseCallbacks` implementation returning `MacroParsingBehavior::Ignore` like:

https://github.com/rust-lang-nursery/rust-bindgen/blob/14da023687b36db480c42f924ec8d236faf1577a/bindgen-integration/build.rs#L21

Or if you don't care about that particular enum either, you could do something like `blacklist_type("IPPORT_.*")`, which should work given it's an anonymous enum I'd think.

**Comment by xxks-kkk** [comment]
Hello @emilio ,

I tried your suggestion and my `build.rs` looks like

```rust
extern crate bindgen;

use std::env;
use std::path::PathBuf;

use bindgen::callbacks::{MacroParsingBehavior, ParseCallbacks};
use std::collections::HashSet;
use std::sync::{Arc, RwLock};


#[derive(Debug)]
struct MacroCallback {
    macros: Arc<RwLock<HashSet<String>>>,
}

impl ParseCallbacks for MacroCallback {
    fn will_parse_macro(&self, name: &str) -> MacroParsingBehavior {
        self.macros.write().unwrap().insert(name.into());

        if name == "IPPORT_RESERVED" {
            return MacroParsingBehavior::Ignore
        }

        MacroParsingBehavior::Default
    }
}

fn main() {
    // The bindgen::Builder is the main entry point
    // to bindgen, and lets you build up options for
    // the resulting bindings.
    let bindings = bindgen::Builder::default()
        .clang_arg("-I/home/zeyuanhu/spdk_install/include")
        // The input header we would like to generate
        // bindings for.
        .header("src/wrapper.h")
        // Finish the builder and generate the bindings.
        .generate()
        // Unwrap the Result and panic on failure.
        .expect("Unable to generate bindings");

    // Write the bindings to the $OUT_DIR/bindings.rs file.
    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

However, I got the following error when I run `cargo build`

```
error[E0432]: unresolved import `bindgen::callbacks::MacroParsingBehavior`
 --> build.rs:6:26
  |
6 | use bindgen::callbacks::{MacroParsingBehavior, ParseCallbacks};
  |                          ^^^^^^^^^^^^^^^^^^^^ no `MacroParsingBehavior` in `callbacks`

error[E0407]: method `will_parse_macro` is not a member of trait `ParseCallbacks`
  --> build.rs:17:5
   |
17 | /     fn will_parse_macro(&self, name: &str) -> MacroParsingBehavior {
18 | |         self.macros.write().unwrap().insert(name.into());
19 | |
20 | |         if name == "IPPORT_RESERVED" {
...  |
24 | |         MacroParsingBehavior::Default
25 | |     }
   | |_____^ not a member of trait `ParseCallbacks`

error: aborting due to 2 previous errors
```

Do you know where I did wrong? Also, can you elaborate more on how to use `blacklist_type("IPPORT_.*")` in the code?

Thanks!

**Comment by xxks-kkk** [comment]
Hello @emilio ,

Thanks for the response. Now, I use `blacklist_type("IPPORT_.*")` inside my `build.rs` but then I run `cargo test`, I got three test cases failure:

```
    bindgen_test_layout_max_align_t
    bindgen_test_layout_spdk_nvme_feat_async_event_configuration
    bindgen_test_layout_spdk_nvme_feat_async_event_configuration__bindgen_ty_1
```

The first test case is probably more related to `bingen` itself (judged by the name). I'm wondering if the failure is related to my `blacklist_type`? What's the test case trying to test? How can I fix it?

Thanks!

**Comment by emilio** [comment]
Hi @xxks-kkk, that makes sense. Those tests failing mean that the struct is unsafe to use from Rust. `max_align_t` is known (#605), and you should just `blacklist_type("max_align_t")` or use `repr(align)` via `rust_target(Stable_1_25)` for now, given otherwise we can't generate a sufficiently aligned type.

The other case looks a bit more problematic. You can mark them as opaque (via `opaque_type`) so bindgen generates the right layout, but that probably deserves a bit of look on its own, we're probably messing up the bitfield calculations. Chances are #1326 could fix it.

Mind filing a new issue about that, with the code snippet you posted plus the definition of `spdk_nvme_critical_warning_state` (and the definitions needed for that and such)? Thinks!

**Comment by emilio** [comment]
Yes, opaque_type will make it sort of unusable inside rust, it'll make it a blob of bytes.

**Comment by stillinbeta** [comment]
For future discoverers of this issue, I'm using this code:

```rust
#[derive(Debug)]
struct IgnoreMacros(HashSet<String>);

impl bindgen::callbacks::ParseCallbacks for IgnoreMacros {
    fn will_parse_macro(&self, name: &str) -> bindgen::callbacks::MacroParsingBehavior {
        if self.0.contains(name) {
            bindgen::callbacks::MacroParsingBehavior::Ignore
        } else {
            bindgen::callbacks::MacroParsingBehavior::Default
        }
    }
}
```

Initialized as 
```rust
let ignored_macros = IgnoreMacros(
            vec![
                "FP_INFINITE".into(),
                "FP_NAN".into(),
                "FP_NORMAL".into(),
                "FP_SUBNORMAL".into(),
                "FP_ZERO".into(),
                "IPPORT_RESERVED".into(),
            ]
            .into_iter()
            .collect(),
        );
```

and called as 
```rust
    .header("wrapper.h")
            .parse_callbacks(Box::new(ignored_macros))
            .rustfmt_bindings(true)
```

**Comment by CommanderTvis** [comment]
Reproduced this issue. What's the workaround of it?
![image](https://user-images.githubusercontent.com/38042667/106353048-4f8ef300-631a-11eb-81a8-b3754737128d.png)

**Comment by db48x** [comment]
Check out @stillinbeta's comment https://github.com/rust-lang/rust-bindgen/issues/687#issuecomment-450750547.

**Comment by CommanderTvis** [comment]
Thank you!

**Comment by CommanderTvis** [comment]
@db48x How to apply that snippet?

**Comment by tgross35** [comment]
Slightly tweaked version, and including `cmath` which defines things like `FE_DIVBYZERO`

```rust
const IGNORE_MACROS
: [&str; 20] = [
    "FE_DIVBYZERO",
    "FE_DOWNWARD",
    "FE_INEXACT",
    "FE_INVALID",
    "FE_OVERFLOW",
    "FE_TONEAREST",
    "FE_TOWARDZERO",
    "FE_UNDERFLOW",
    "FE_UPWARD",
    "FP_INFINITE",
    "FP_INT_DOWNWARD",
    "FP_INT_TONEAREST",
    "FP_INT_TONEARESTFROMZERO",
    "FP_INT_TOWARDZERO",
    "FP_INT_UPWARD",
    "FP_NAN",
    "FP_NORMAL",
    "FP_SUBNORMAL",
    "FP_ZERO",
    "IPPORT_RESERVED",
];

#[derive(Debug)]
struct IgnoreMacros(HashSet<String>);

impl ParseCallbacks for IgnoreMacros {
    fn will_parse_macro(&self, name: &str) -> MacroParsingBehavior {
        if self.0.contains(name) {
            MacroParsingBehavior::Ignore
        } else {
            MacroParsingBehavior::Default
        }
    }
}

impl IgnoreMacros {
    fn new() -> Self {
        Self(IGNORE_MACROS
            .into_iter().map(|s| s.to_owned()).collect())
    }
}

```

And usage

```rust
.parse_callbacks(Box::new(IgnoreMacros::new()))
// Longdouble creates warnings about 16-bit numbers, good to ignore if relevant
.blocklist_function("strtold")
```

I wonder if `bindgen` could ignore these specific macros by default with the option to disable them, that would save some head scratching.

**Comment by wildwestrom** [comment]
I see this issue is labeled as fix, but it must have regressed. I used the same header and bindgen, and it gave me almost the exact same error. I can post more details if necessary.

**Comment by pvdrz** [comment]
This issue has not regressed, what happens is that there is no magic solution for it. As mentioned [https://github.com/rust-lang/rust-bindgen/issues/687#issuecomment-416537395](before), you need to either block the macro or the enumerator. A more in-depth explanation is given [here](https://github.com/rust-lang/rust-bindgen/issues/2596#issuecomment-1670143265)
